include/kalmar_math.h

```diff
@@ -486,10 +486,10 @@ namespace Kalmar
         float fmod(float x, float y) { return fast_math::fmodf(x, y); }
 
         HCC_MATH_LIB_FN
-        float frexpf(float x, int *exp) { return __ocml_frexp_f32(x, exp); }
+        float frexpf(float x, int *exp) { return __ocml_frexp_f32(x, (__attribute__((address_space(5))) int*)exp); }
```

Comment:
second definition in line 1163

Reply:
the one at 1163 is in the kalmar::precise_math namespace.  We also have a kalmar::fast_math namespace.  I don't know what the original requirements were for creating these namespaces, but they're there.  Originally, both would just call the __hc math functions.  Now, we're calling the ocml functions directly, but I left the fast_math and precise_math namespaces.  Maybe we don't need them anymore?  But I don't know if there are any dependencies on them in ROCm.

Reply:
They are there by design, do not remove. Thanks.

Reply:
ok.  Thanks Alex.

Reply:
Okay lets keep it as it is. Do we need ocml functions fract, lgamma_r, and sincospi?

Reply:
yeah, it seems we've missed these math functions.  There may be others; I'd have to go through all of the ocml math functions to check.  Unless someone knows of a reason why we excluded these?

Reply:
I have a feeling those missing math functions were added to ocml after Kalmar worked on these in HCC. Similar to bitselect, bitinsert, and bitextract which I added in Aug.

Reply:
Thanks Aaron.  I'll add these.  I need to correct the math functions with pass-by-ref args as well.  See comment above.

Reply:
ok ... correction to my above comment.  the bit* functions you mentioned above are builtins - which is a bit lower than these math RT functions, which may call the builtins.  As for fract, lgamma_r, and sincospi; we could add these, but there are a bunch of other OCML functions we're also missing.  Does anyone have objections to adding more than just those 3?

Reply:
for the record, this is the list of OCML math functions not wrapped in this header:
acospi
add
asinpi
atanpi
atan2pi
atanpired
atanred
erfcx
fmuladd
fract
len3
len4
lgamma_r
mad
max
maxmag
min
minmag
mul
ncdf
ncdfinv
pown
powr
rhypot
rint
rootn
sincospi
sincospired
sub
tanpired
tanred
tgamma

Reply:
Thanks @david-salinas, that's a lot. Please take your time. also I do not disagree with adding these.

Reply:
@AlexVlx @b-sumner Do either of you know why we seem to avoid double (f64) floats when we wrap the ocml math functions for HC users?  Is f64 support on accelerators frowned upon?  Should I add them?

Reply:
My answers are: No idea.  No.  Yes.

Reply:
No you should not. By design fast_math does not expose doubles, only floats (and FP16). No, you should not add doubles into that namespace. No you should not add all functions in ROCDL, only if you identify some function that is in the {C++ standard library, C++AMP} union but is missing here should we add them. If both of you look, precise_math has lots and lots of doubles. This is intentional and not an accident.

Reply:
I think that we should be matching C++11 / C++AMP (without some functions in AMP 1.2, which were... special). It was the case way back when I checked / touched these. Having said that, we may have some extra barnacles due to HIP, which we may be able to clean now that HIP moved on to do the right thing i.e. consume ROCDL directly.

Reply:
ok, I may not be looking at the official C++AMP amp_math.h, but as far as I can tell, from my list above only, rint and tgamma are in amp_math.h and C++ stdlib, and are missing here.

Reply:
`tgamma` not in fast_math. fast_math != precise_math in terms of the functions that are exposed. Do not look at headers that were perhaps hacked for reasons lost to history, look at official specs: http://download.microsoft.com/download/2/2/9/22972859-15C2-4D96-97AE-93344241D56C/CppAMPOpenSpecificationV12.pdf. `rint` is more subtle since it needs to access the current rounding mode to meet the spec: https://en.cppreference.com/w/cpp/numeric/math/rint, which is possibly problematic. It was intentionally omitted from AMP, but we may want to consider adding it in the future.

Reply:
ok.  So we really don't need to add any then.  Looks like we're covered.

Reply:
updated commit pushed